src/arch/arm/64/kernel/vspace.c

```diff
@@ -1622,7 +1664,18 @@ static exception_t decodeARMFrameInvocation(word_t invLabel, word_t length,
             return EXCEPTION_SYSCALL_ERROR;
         }
 #endif
-
+#ifndef CONFIG_ARM_HYPERVISOR_SUPPORT
+        /* When in EL1, the mapping must be write-able for ARMPageInvalidate_Data */
+        if (invLabel == ARMPageInvalidate_Data) {
+            lookupPTSlot_ret_t lu_ret = lookupPTSlot(find_ret.vspace_root, vaddr);
+            vm_rights_t vm_rights = vmRightsFromPTE(*lu_ret.ptSlot);
```

Comment:
It has to be a page at this point, because we're only calling this from a mapped page cap (the decode function bails out above if it's not mapped, and the cap type makes sure it's page).

Reply:
I am indeed wrong, the cap mapping info could be out of date, and we could at this stage also get an invalid entry, not just a page type entry. Will add a check for that, and I agree that it should silently ignore it as in the VSpace case.

Reply:
Hm, it's even worse. We could actually be getting a completely different page -- the invariants only give guarantees about mapping info for page tables, not for pages (this is part of what this caveat in #1277 is about). So I think I need to check the frame pointer as well, like in `unmapPage`.

Reply:
You're right, the `asid` check assures the page was mapped. The only issue is the cap data being possibly out-of-date.

>  I agree that it should silently ignore it as in the VSpace case.

In the VSpace case it's silently ignored because there is no mapping and hence nothing to do. I think it's fine to return an error in this case, because the user is using a stale cap, which is unexpected.

Any reason why frame caps weren't made children of the PT cap to avoid these kind of issues? To me that seems easier for verification than all the fixups.

Reply:
> Any reason why frame caps weren't made children of the PT cap to avoid these kind of issues? To me that seems easier for verification than all the fixups.

Never mind, I forgot the obvious reason that there is a memory frame bound to the page cap and that's why it needs to stay a child of the parent untyped.